plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseClient.java

```diff
@@ -842,4 +871,229 @@ private static SliceWriteFunction uuidWriteFunction()
     {
         return (statement, index, value) -> statement.setObject(index, trinoUuidToJavaUuid(value), Types.OTHER);
     }
+
+    // Different versions of ClickHouse may support different min/max values for the
+    // same data type, you can refer to the table below:
+    //
+    // | version | column type | min value           | max value            |
+    // |---------|-------------|---------------------|----------------------|
+    // | any     | UInt8       | 0                   | 255                  |
+    // | any     | UInt16      | 0                   | 65535                |
+    // | any     | UInt32      | 0                   | 4294967295           |
+    // | any     | UInt64      | 0                   | 18446744073709551615 |
+    // | < 21.4  | Date        | 1970-01-01          | 2106-02-07           |
+    // | < 21.4  | DateTime    | 1970-01-01 00:00:00 | 2106-02-06 06:28:15  |
+    // | >= 21.4 | Date        | 1970-01-01          | 2149-06-06           |
+    // | >= 21.4 | DateTime    | 1970-01-01 00:00:00 | 2106-02-07 06:28:15  |
+    //
+    // And when the value written to ClickHouse is out of range, ClickHouse will store
+    // the incorrect result, so we need to check the range of the written value to
+    // prevent ClickHouse from storing the incorrect value.
+    private enum TrinoToClickHouseWriteChecker
```

Comment:
Should this be a separate class - it's a lot of logic and we'd probably want to share some things from here in tests.